paddle/fluid/memory/stats.h

```diff
@@ -0,0 +1,176 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <atomic>
+#include <map>
+#include <string>
+#include "paddle/fluid/framework/new_executor/workqueue/thread_data_registry.h"
+#include "paddle/fluid/platform/enforce.h"
+#include "paddle/fluid/platform/errors.h"
+#include "paddle/fluid/platform/macros.h"
+
+namespace paddle {
+namespace memory {
+
+using framework::ThreadDataRegistry;
+
+struct ThreadLocalStatBase {
+  int64_t current{0};
+  int64_t peak{0};
+};
+
+class StatBase {
+ public:
+  StatBase() = default;
+  virtual ~StatBase() = default;
+
+  virtual int64_t GetCurrentValue() = 0;
+  virtual int64_t GetPeakValue() = 0;
+  virtual void Update(int64_t) = 0;
+
+ private:
+  DISABLE_COPY_AND_ASSIGN(StatBase);
+};
+
+template <typename ThreadLocalStatType>
+class Stat : public StatBase {
+ public:
+  static Stat* GetInstance() {
+    static Stat instance;
+    return &instance;
+  }
+
+  int64_t GetCurrentValue() override {
+    std::unordered_map<uint64_t, ThreadLocalStatType> thread_local_stats =
+        ThreadDataRegistry<ThreadLocalStatType>::GetInstance()
+            .GetAllThreadDataByValue();
+    int64_t current_value = 0;
+    for (auto pair : thread_local_stats) {
+      current_value += pair.second.current;
+    }
+    return current_value;
+  }
+
+  int64_t GetPeakValue() override { return peak_value_; }
+
+  void Update(int64_t increment) override {
+    ThreadLocalStatType thread_local_stat =
+        ThreadDataRegistry<ThreadLocalStatType>::GetInstance()
+            .GetCurrentThreadData();
```

Comment:
GetMutableCurrentThreadData is designed for read&write scene

Reply:
done, thx.